public/app/plugins/datasource/loki/querySplitting.ts

```diff
@@ -0,0 +1,91 @@
+import { Subscriber, map, Observable } from 'rxjs';
+
+import { DataQueryRequest, DataQueryResponse, dateTime, TimeRange } from '@grafana/data';
+import { LoadingState } from '@grafana/schema';
+
+import { LokiDatasource } from './datasource';
+import { getRanges } from './metricTimeSplit';
+import { combineResponses, resultLimitReached } from './queryUtils';
+import { LokiQuery } from './types';
+
+function partitionTimeRange(originalTimeRange: TimeRange, intervalMs: number, resolution: number): TimeRange[] {
+  // we currently assume we are only running metric queries here.
+  // for logs-queries we will have to use a different time-range-split algorithm.
+
+  // the `step` value that will be finally sent to Loki is rougly the same as `intervalMs`,
+  // but there are some complications.
+  // we need to replicate this algo:
+  //
+  // https://github.com/grafana/grafana/blob/main/pkg/tsdb/loki/step.go#L23
+
+  const start = originalTimeRange.from.toDate().getTime();
+  const end = originalTimeRange.to.toDate().getTime();
+
+  const safeStep = Math.ceil((end - start) / 11000);
```

Comment:
😄 
the short answer is that we are duplicating the go-code from https://github.com/grafana/grafana/blob/main/pkg/tsdb/loki/step.go#L23

the long answer:
there is a limit in Loki (well, there is one in Prometheus, and i think Loki has that too), the max number of timestamps with data it is willing to return is 11000.

so the code has to ensure that the request we send will not cause more than 11000 results.

so we calculate what kind of `step` value would cause the query to return 11000 results (the `safeStep`), and then make sure the calculated step is not smaller. (smaller-step makes more-timestamps)